init/lib/status.go

```diff
@@ -17,38 +17,111 @@ package lib
-	pidBytes, err := ioutil.ReadFile(Pidfile)
+type ServicePids struct {
+	PidsByName map[string]int `yaml:"pidsByName" validate:"nonzero"`
+}
+
+type CmdWithOutputFile struct {
+	Cmd            *exec.Cmd
+	OutputFilename string
+}
+
+func GetNotRunningCmdsByName() (map[string]CmdWithOutputFile, error) {
+	cmdsByName, err := GetConfiguredCommandsByName()
 	if err != nil {
-		return nil, 3, errors.Wrap(err, "failed to read pidfile")
+		return nil, errors.Wrap(err, "failed to get commands from static and custom configuration files")
 	}
+	runningProcsByName, err := GetRunningProcsByName()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to determine running processes")
+	}
+	notRunningCmdsByName := make(map[string]CmdWithOutputFile)
+	for name, cmd := range cmdsByName {
+		if _, ok := runningProcsByName[name]; !ok {
+			notRunningCmdsByName[name] = cmd
+		}
+	}
+	return notRunningCmdsByName, nil
+}
 
-	pid, err := strconv.Atoi(string(pidBytes[:]))
+func GetRunningProcsByName() (map[string]*os.Process, error) {
+	pidfileBytes, err := ioutil.ReadFile(pidfile)
 	if err != nil {
-		return nil, 3, errors.Wrap(err, "failed to parse a valid PID from pidfile")
+		return nil, errors.Wrap(err, "failed to read pidfile")
+	}
+	var servicePids ServicePids
+	if err := yaml.Unmarshal(pidfileBytes, &servicePids); err != nil {
+		return nil, errors.Wrap(err, "failed to deserialize pidfile")
+	}
+	if err := validator.Validate(servicePids); err != nil {
+		return nil, errors.Wrap(err, "failed to deserialize pidfile")
 	}
 
-	// Docs say FindProcess always succeeds on Unix.
-	process, _ := os.FindProcess(pid)
-	if !isRunning(process) {
-		return process, 1, errors.New("pidfile exists but process is not running")
+	runningProcs := make(map[string]*os.Process)
+	for name, pid := range servicePids.PidsByName {
+		running, proc := isPidRunning(pid)
+		if running {
+			runningProcs[name] = proc
+		}
 	}
+	return runningProcs, nil
+}
 
-	return process, 0, nil
+func GetConfiguredCommandsByName() (cmdsByName map[string]CmdWithOutputFile, rErr error) {
+	primaryStdout, err := os.Create(launchlib.PrimaryOutputFile)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create primary output file: "+launchlib.PrimaryOutputFile)
+	}
+	defer func() {
+		if cErr := primaryStdout.Close(); rErr == nil && cErr != nil {
+			rErr = errors.Wrap(err, "failed to close primary output file: "+launchlib.PrimaryOutputFile)
+		}
+	}()
+	staticConfig, customConfig, err := launchlib.GetConfigsFromFiles(launcherStaticFile, launcherCustomFile,
+		primaryStdout)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to read static and custom configuration files")
+	}
+	serviceCmds, err := launchlib.CompileCmdsFromConfig(&staticConfig, &customConfig, primaryStdout)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to compile commands from static and custom configurations")
+	}
+	cmdsByName = make(map[string]CmdWithOutputFile)
+	cmdsByName["primary"] = CmdWithOutputFile{
```

Comment:
This is still suffering from the preconceived notion that "all processes should be treated exactly the same" when we do not in fact treat them the same.  The config defines them as different, the log output defines them as different, and we want users to think of them as being different.  I don't understand why we are then constantly trying to shoehorn that into this model of everything being treated the same?

Not only does it introduce code smells (of still using a string to define a concept), but it is fundamentally buggy, since I can now define a valid configuration which will not work with `go-init`.